waspls/src/Wasp/LSP/Server.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExplicitNamespaces #-}
+
+module Wasp.LSP.Server
+  ( serve,
+  )
+where
+
+import qualified Control.Concurrent.MVar as MVar
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import qualified Control.Monad.Trans.Except as Except
+import qualified Control.Monad.Trans.State.Strict as State
+import qualified Data.Aeson as Aeson
+import Data.Default (Default (def))
+import qualified Data.Text as Text
+import qualified Language.LSP.Server as LSP
+import qualified Language.LSP.Types as LSP
+import System.Exit (ExitCode (ExitFailure), exitWith)
+import qualified System.Log.Logger
+import Wasp.LSP.Core (ServerConfig, ServerError (ServerError), ServerM, ServerState, Severity (..))
+import Wasp.LSP.Handlers
+
+serve :: Maybe FilePath -> IO ()
+serve maybeLogFile = do
+  setupLspLogger maybeLogFile
+
+  let defaultServerState = def :: ServerState
+  state <- MVar.newMVar defaultServerState
+
+  let lspServerInterpretHandler env =
+        LSP.Iso {forward = runHandler, backward = liftIO}
+        where
+          runHandler :: ServerM a -> IO a
+          runHandler handler =
+            -- Get the state from the "MVar", run the handler in IO and update
+            -- the "MVar" state with the end state of the handler.
+            MVar.modifyMVar state \oldState -> do
+              LSP.runLspT env do
+                (e, newState) <- State.runStateT (Except.runExceptT handler) oldState
+                result <- case e of
+                  Left (ServerError severity errMessage) -> sendErrorMessage severity errMessage
+                  Right a -> return a
+
+                return (newState, result)
+
+  exitCode <-
+    LSP.runServer
+      LSP.ServerDefinition
+        { defaultConfig = def :: ServerConfig,
+          onConfigurationChange = lspServerUpdateConfig,
+          doInitialize = lspServerDoInitialize,
+          staticHandlers = lspServerHandlers,
+          interpretHandler = lspServerInterpretHandler,
+          options = lspServerOptions
+        }
+
+  case exitCode of
+    0 -> return ()
+    n -> exitWith (ExitFailure n)
+
+-- | Setup global DEBUG logger. Logs at other levels are ignored.
+--
+-- Use 'System.Log.Logger.logM' at "DEBUG" level to write to this log.
+--
+-- @setupLspLogger Nothing@ doesn't set up any logger, so logs are not output
+-- anywhere.
+--
+-- @setupLspLogger (Just "[OUTPUT]")@ sends log messages to the LSP client
+--
+-- @setupLspLogger (Just filepath)@ writes log messages to the path given
+setupLspLogger :: Maybe FilePath -> IO ()
+setupLspLogger Nothing = pure ()
+setupLspLogger (Just "[OUTPUT]") = LSP.setupLogger Nothing [] System.Log.Logger.DEBUG
+setupLspLogger file = LSP.setupLogger file [] System.Log.Logger.DEBUG
+
+lspServerUpdateConfig :: ServerConfig -> Aeson.Value -> Either Text.Text ServerConfig
```

Comment:
We could write a comment saying that the Text in Either is an Json parsing error. Because we can't figure that really out from the name of the function or from the type signature.